File: javadarts/src/main/java/com/sdoward/barstewarddarts/android/java/Game.java
package com.sdoward.barstewarddarts.android.java;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.Iterator;
import java.util.List;
import java.util.Map;

public class Game {

    private final GameInstructions gameInstructions;
    private final List<Player> players;
    private final List<Turn> turns = new ArrayList(100);
    private final TurnValidator turnValidator = new TurnValidator();

    private int pointer = 0;

    public Game(List<Player> players, GameInstructions gameInstructions) {
        this.players = players;
        this.gameInstructions = gameInstructions;
    }

    public void start() {
        displayCurrentPlayer();
        pointer++;
    }

    public List<Player> getPlayers() {
        return players;
    }

    public void acceptThrow(Throw thro) {
        turnValidator.acceptThrow(thro);
        if (turnValidator.turnIsComplete()) {
            acceptTurn(turnValidator.getTurn());
        }
    }

    private void acceptTurn(Turn turn) {
        reconsileLives(turn.getFirstThrow());
        reconsileLives(turn.getSecondThrow());
        reconsileLives(turn.getThirdThrow());
        turns.add(turn);
        displayDrinkingInstructions(turn);
        displayCurrentPlayer();
        if (hasGameEnded()) {
            gameInstructions.displayWinner(players.get(pointer).getName());
        }
        if (pointer == players.size() - 1) {
            pointer = 0;
        } else {
            pointer++;
        }
    }

    private void displayCurrentPlayer() {
        gameInstructions.displayCurrentPlayer(players.get(pointer).getName());
    }

    private void reconsileLives(Throw thro) {
        for (int i = 0; i < players.size(); i++) {
            Player player = players.get(i);
            if (i == pointer) {
                if (thro.getNumber() == player.getNumber())
                    player.setLives(player.getLives() - thro.getLifeValue());
            } else {
                if (thro.getNumber() == player.getNumber())
                    player.setLives(player.getLives() + thro.getLifeValue());
            }
        }
    }

    private void displayDrinkingInstructions(Turn turn) {
        Map<Player, Integer> drinkers = new HashMap();
        for (int i = 0; i < players.size(); i++) {
            Player player = players.get(i);
            addDrinking(drinkers, turn.getFirstThrow(), player);
            addDrinking(drinkers, turn.getSecondThrow(), player);
            addDrinking(drinkers, turn.getThirdThrow(), player);
        }
        List<String> drinkingInstructions = new ArrayList(drinkers.size());
        for (Map.Entry<Player, Integer> entry: drinkers.entrySet()) {
            drinkingInstructions.add(entry.getKey().getName() + " drink " + entry.getValue());
        }
        gameInstructions.displayDrinkingInstructions(drinkingInstructions);
    }

    private void addDrinking(Map<Player, Integer> drinkers, Throw thro, Player player) {
        if (thro.getNumber() == player.getNumber()) {
            if (drinkers.containsKey(player)) {
                int newDrinkValue = drinkers.get(player) + thro.getLifeValue();
                drinkers.remove(player);
                drinkers.put(player, newDrinkValue);
            } else {
                drinkers.put(player, thro.getLifeValue());
            }
        }
    }

    private boolean hasGameEnded() {
        for (Player player : players) {
            if (player.getLives() < 1)
                return true;
        }
        return false;
    }

    interface GameInstructions {

        void displayCurrentPlayer(String name);

        void displayWinner(String name);

        void displayDrinkingInstructions(List<String> drinkers);

    }

}

File: kotlindarts/src/main/kotlin/com/sdoward/barstewarddarts/android/kotlin/Game.kt
package com.sdoward.barstewarddarts.android.kotlin

class Game(val players: List<Player>, private val gameInstructions: GameInstructions) {

    private val turns = mutableListOf<Turn>()

    private var pointer = 0

    fun start() {
        displayCurrentPlayer()
        pointer++
    }

    fun acceptThrow(thro: Throw) {
        TurnValidator.acceptThrow(thro)
        if (TurnValidator.turnIsComplete()) {
            acceptTurn(TurnValidator.turn)
        }
    }

    private fun acceptTurn(turn: Turn) {
        turn.toList().forEach { reconsileLives(it) }
        turns.add(turn)
        displayDrinkingInstructions(turn)
        displayCurrentPlayer()
        if (players.any { it.lives < 1 }) {
            gameInstructions.displayWinner(players[pointer].name)
        }
        if (pointer == players.size - 1) {
            pointer = 0
        } else {
            pointer.inc()
        }
    }

    private fun displayCurrentPlayer() {
        gameInstructions.displayCurrentPlayer(players[pointer].name)
    }

    private fun reconsileLives(thro: Throw) {
        for (i in players.indices) {
            val player = players[i]
            if (numberHit(thro, player)) {
                if (i == pointer) {
                    player.lives = player.lives - thro.lifeValue
                } else {
                    player.lives = player.lives + thro.lifeValue
                }
            }
        }
    }

    private fun displayDrinkingInstructions(turn: Turn) {
        players.filter {
            turn.toList().any {thro -> numberHit(thro, it) }
        }.map {
            val drinkAmount = turn.toList().fold(0) {
                total, thro -> total.plus(getDrinkAmount(thro, it))
            }
            it.name.plus(" drink ").plus(drinkAmount)
        }.let {
            gameInstructions.displayDrinkingInstructions(it)
        }
    }

    private fun getDrinkAmount(thro: Throw, player: Player): Int {
        if (numberHit(thro, player)) {
            return thro.lifeValue
        } else {
            return 0
        }
    }

    private fun numberHit(thro: Throw, player: Player): Boolean {
        return thro.number == player.number
    }

}

interface GameInstructions {

    fun displayCurrentPlayer(name: String)

    fun displayWinner(name: String)

    fun displayDrinkingInstructions(drinkers: List<String>)

}
